File: leetcode/0155. Min Stack/Solution.swift
// https://leetcode.com/problems/min-stack/

class MinStack {

    typealias Element = (value: Int, minimum: Int)

    private var stack = [Element]()

    func push(_ x: Int) {
        stack.append(Element(value: x, minimum: stack.isEmpty ? x : min(x, getMin())))
    }

    func pop() {
        _ = stack.removeLast()
    }

    func top() -> Int {
        return stack.last?.value ?? Int.min
    }

    func getMin() -> Int {
        return stack.last?.minimum ?? Int.min
    }
}

class MinStack<T: Numeric & Comparable> {

    typealias Element = (value: T, minimum: T)

    private var stack = [Element]()

    func push(_ x: T) {
        stack.append(Element(value: x, minimum: stack.isEmpty ? x : min(x, getMin())))
    }

    func pop() {
        _ = stack.removeLast()
    }

    func top() -> T {
        return stack.last?.value ?? T.zero
    }

    func getMin() -> T {
        return stack.last?.minimum ?? T.zero
    }
}

let s = MinStack<Int>()
s.push(-2)
s.push(0)
s.push(-3)
assert(s.getMin() == -3)
s.pop()
assert(s.top() == 0)
assert(s.getMin() == -2)

File: leetcode/0155. Min Stack/Solution.py
# https://leetcode.com/problems/min-stack/

class MinStack:

    def __init__(self):
        self.stack = []
        

    def push(self, x: int) -> None:
        min_value = 0
        if not self.stack:
            min_value = x
        else:
            min_value = min(x, self.getMin())
        self.stack.append((x, min_value))
        

    def pop(self) -> None:
        if self.stack:
            self.stack.pop()


    def top(self) -> int:
        if self.stack:
            (v, m) = self.stack[-1]
            return v
        else:
            return 0
        

    def getMin(self) -> int:
        if self.stack:
            (v, m) = self.stack[-1]
            return m
        else:
            return 0

min_stack = MinStack()
min_stack.push(-2)
min_stack.push(0)
min_stack.push(-3)
assert min_stack.getMin() == -3
min_stack.pop()
assert min_stack.top() == 0
assert min_stack.getMin() == -2
